test/cql-pytest/test_wasm.py

```diff
@@ -0,0 +1,317 @@
+# -*- coding: utf-8 -*-
+# Copyright 2021-present ScyllaDB
+#
+# This file is part of Scylla.
+#
+# Scylla is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
```

Comment:
Need to add a test for infinite loops, to see they terminate, and measure the time it takes.

Reply:
There already is such a test case for a really large fibonacci number - which correctly fails due to not enough fuel

Reply:
But I'll add a specific case for an infinite loop and measure the result

Reply:
Done. I measured inside Python, so some overhead may be involved, but the results were generally as follows:
```bash
sarna@srn scylla]$ for i in {1..10}; do pytest -s -vvv test/cql-pytest/test_wasm.py::test_infinite_loop |& grep loop\ took; done
test/cql-pytest/test_wasm.py::test_infinite_loop Breaking the loop took 2.69ms
test/cql-pytest/test_wasm.py::test_infinite_loop Breaking the loop took 2.79ms
test/cql-pytest/test_wasm.py::test_infinite_loop Breaking the loop took 2.59ms
test/cql-pytest/test_wasm.py::test_infinite_loop Breaking the loop took 2.50ms
test/cql-pytest/test_wasm.py::test_infinite_loop Breaking the loop took 2.70ms
test/cql-pytest/test_wasm.py::test_infinite_loop Breaking the loop took 2.81ms
test/cql-pytest/test_wasm.py::test_infinite_loop Breaking the loop took 2.71ms
test/cql-pytest/test_wasm.py::test_infinite_loop Breaking the loop took 2.68ms
test/cql-pytest/test_wasm.py::test_infinite_loop Breaking the loop took 2.94ms
test/cql-pytest/test_wasm.py::test_infinite_loop Breaking the loop took 2.84ms

```

Reply:
We need to go below 0.5ms (and then some, to account for slower machines typically used in servers).

Reply:
Unsurprisingly, time measurements performed with Python turned out to be absolutely useless, so I just injected a timer directly into the code. For 200k units of fuel, I got consistent results of execution times below 0.4ms. For 256k units I occasionally saw the time to exceed 0.5ms, so 200k is picked as the first lower number that looks nice as a constant.

Note that compiling the script can take longer - I saw ~1ms even for these simple functions, but there's not so much we can do about it, except for assuming that creating a function and propagating it to other nodes is a rare operation.

Reply:
That translates to 1 fuel = 2ns. But we need to go lower. First, your machine is likely clocked higher than a busy server. Second, the server shard can compete with another hyperthread. Third, we don't want to be close to 0.5ms, since we can start execution close to the preemption point. We need to target around 0.1ms. Assuming factors of (2, 2, 4) for these, we get 12.5k. Let's round it down to 10k for extra safety.

Reply:
(also, an infinite loop is likely executed very quickly compared to branchy and memory-accessing code)

Reply:
Lowering to 10k makes one of my tests fail - the one which recursively (and naively) computes 14th Fibonacci number. Bumping to 20k is already enough fuel to make the function pass. I know that it's an absolutely random point of reference, but perhaps we can settle on 20k for starters? Otherwise I'll just degrade the test to a lower number, but on the other hand, 14th Fibonacci doesn't sound that CPU-intensive.

Reply:
By the way - wasmer from wasmer.io has a concept of different fuel prices per wasm operation: https://docs.rs/wasmer-c-api/2.0.0/wasmer_c_api/wasm_c_api/unstable/middlewares/metering/index.html

I haven't found matching wasmtime code yet, but perhaps they use weighted fuel prices as well, which could mean that an infinite loop also uses less fuel per op than branching.

Reply:
update: the fuel prices aren't that complicated, and branching is actually totally disregarded:
```rust

        self.fuel_consumed += match op {
            // Nop and drop generate no code, so don't consume fuel for them.
            Operator::Nop | Operator::Drop => 0,

            // Control flow may create branches, but is generally cheap and
            // free, so don't consume fuel. Note the lack of `if` since some
            // cost is incurred with the conditional check.
            Operator::Block { .. }
            | Operator::Loop { .. }
            | Operator::Unreachable
            | Operator::Return
            | Operator::Else
            | Operator::End => 0,

            // everything else, just call it one operation.
            _ => 1,
        };
```
src: https://github.com/bytecodealliance/wasmtime/blob/63a3bbbf5ad29f194480aa4064ba431390e70050/crates/cranelift/src/func_environ.rs#L371-L387